### PR TITLE
Refactor `elasticsearch` module for modular and extensible filter handling

### PR DESCRIPTION
**Pull request by cewald**
In order to the PR https://github.com/DivanteLtd/storefront-query-builder/pull/4#issuecomment-609625727 and #3, we decided to give a more extensible approach in adding features to the `storefront-query-builder` by doing some refactoring.

I refactored the `src/elasticsearch/index.ts` and put the logic into a class to make it more read- and extendable. I also added the following features:
* Add `applySort` method to `SearchQuery` to add more complex, custom sorts from `SearchQuery`
* Add `customFilters` option to `buildQueryBodyFromSearchQuery` to be able to extend `SearchQuery` by custom filters
* Add `or`, `nor` and `nin` filters to `baseFilters`
* Refactor `buildQueryBodyFromSearchQuery` into more modular class – The goal was to have more readable code and be able to extend the base filter functions in a more logical way
* Added some `eslint` and `typescript` improvements for better coding experience

There are some pitfalls and thoughts about backwards compatibility to do yet:
* The unit tests of the VSF will fail with this despite the new `_applySort` property in `SearchQuery` – this is really easy to fix, but would appear in each test running against old code
* The API would need an extension to somehow support the new `customFilters` catalog-endpoint-wise without a need to overwrite this endpoint
* The `api-search-query` adapter in the Vue Storefront will need some small updates for the new sorting options

So it would be a good idea to manage this changes under a new version number to support older systems by the current version and just use this version from the current VSF/API `develop` branches on as a new version.

This is currently in progress and not much tested – but please give me some feedback if I don't head in a completely wrong direction.

### CODE REVIEW

File: src/elasticsearch/index.ts [hunk 1]
```diff
@@ -1,166 +1,440 @@
...
- *
- * @param {Object} object
- * @param {String} scope
- * @returns {boolean}
+ * Class to create filter in modular and reproductive way
  */
-function checkIfObjectHasScope ({ object, scope }: { object: { scope: string }, scope: string}) {
-  return object.scope === scope || (Array.isArray(object.scope) && object.scope.indexOf(scope) >= 0);
-}
+export default class Filters {
```
Comment (gibkigonzo) [inline src/elasticsearch/index.ts:30]: what do you think about moving this into separete file?

File: src/elasticsearch/index.ts [hunk 2]
```diff
@@ -1,166 +1,440 @@
...
-  }
-  return processedFilter;
+export interface FilterOptions {
+  operator: string
+  attribute: string
+  value: any,
+  queryChain?: any
+}
+
+export interface FilterInterface {
```
Comment (gibkigonzo) [inline src/elasticsearch/index.ts:16]: Can you please describe all those methods as much as possible? When someone will import this interface to create customFilters then he will know what he need to do

File: src/elasticsearch/index.ts [hunk 3]
```diff
@@ -1,166 +1,440 @@
...
+        return
+      }
+
+      value = typeof value === 'object' ? value : { 'in': [ value ] }
+      if (Object.keys(value).length > 0) {
+        attribute = this.getMapping(attribute)
+        const operator = Object.keys(value)[0]
+
+        this.getSortedFilters().forEach(filterHandler => {
+          // Prevent duplicate assign of filters, despite the fact you can't break `forEach`
```
Comment (gibkigonzo) [inline src/elasticsearch/index.ts:196]: how about using `for in` loop? :thinking: then you can return from it and clear this `filterApplied`

File: src/elasticsearch/index.ts (hunk 2 above)
Comment (cewald) [reply src/elasticsearch/index.ts:16]: I will using JSDocs. But actually I also would write a recipe in the VSF docs as well.

File: src/elasticsearch/index.ts (hunk 3 above)
Comment (cewald) [reply src/elasticsearch/index.ts]: Done

File: src/elasticsearch/index.ts (hunk 2 above)
Comment (cewald) [reply src/elasticsearch/index.ts]: If described the main functions and interfaces as much as possible.

File: src/elasticsearch/index.ts (hunk 1 above)
Comment (cewald) [reply src/elasticsearch/index.ts]: Done

**Comment by cewald** [comment]
I have tested it a bit now and it seems to work fine.
I needed to make some changes at the API and VSF to make it work.

Changes for `vue-storefront`:
* Update the Unit-Tests to accept the `SearchQuery` object
* Connect new branch or `npm` version of `storefront-query-builder`
* Update Docs – add a recipe on how to add a custom filter

Changes for `vue-storefront-api`:
* I've added a helper that searches for specific filter classes in enabled extension folders and adds them to the catalog endpoint – this could be easily implemented on other ES queries as well this way
* Connect new branch or `npm` version of `storefront-query-builder`

I will open up two PR's in the repos and connect this ticket.

**Comment by gibkigonzo** [comment]
Thanks, this looks great! :)

**Pull request merged**